### PR TITLE
fix(web-shell): enable Changes and History dialogs for worktree sessions

### PR DESCRIPTION
**Pull request by wenshao**
The Web Shell's Changes (diff) and History (log) dialogs were completely disabled for worktree sessions — the composer button was hidden and the `/diff` / `/log` slash commands always targeted the main workspace root. This PR threads an optional git working directory through the entire stack (daemon routes → SDK client → dialog components) so that worktree sessions can view their own file changes and commit history, matching the existing git-status chip behavior that already supported worktrees.

The path containment validation logic (realpath + relative-path check) that was previously inlined in the git status route is extracted into a shared helper and reused by the diff and log routes.

## Why it's needed

When a session runs inside a git worktree, users could see the branch indicator (purple fork icon with dirty/ahead/behind counts) but could not open the Changes or History dialog to inspect what actually changed. The only way to see worktree diffs was through the `exit_worktree` tool output — not accessible from the UI. This made worktree sessions feel second-class compared to regular sessions.

### How to verify

1. Start `qwen serve` and open the Web Shell.
2. Create a worktree session (select "Worktree" mode in the git popover before sending the first message).
3. Make some file changes inside the session (ask the agent to edit a file).
4. Click the git chip in the composer toolbar — the Changes dialog should open and show the worktree's modified files with correct diffs.
5. Switch to the History tab — it should show the worktree branch's commit log.
6. Verify `/diff` and `/log` slash commands also open the dialog targeting the worktree.
7. Confirm non-worktree sessions still work as before (Changes/History target the workspace root).

### Evidence (Before & After)

Before: worktree sessions had no Changes button (hidden by `!sessionWorktree` guard); `/diff` and `/log` targeted the main workspace root.
After: worktree sessions show the Changes button; dialogs target the worktree path via `?cwd=` query parameter.

### Tested on

|     OS     | Status |
| :--------: | :----: |
|  🍏 macOS  |   ✅   |
| 🪟 Windows |   ⚠️   |
|  🐧 Linux  |   ⚠️   |

### Environment (optional)

Unit tests: CLI route tests (33 passed), SDK DaemonClient tests (292 passed), Web Shell dialog tests (26 passed). Typecheck clean on all three packages (no new errors).

## Risk & Scope

- Main risk or tradeoff: the `?cwd=` parameter accepts an arbitrary path from the client, but it is validated by realpath resolution and containment within the workspace root (same pattern already used by the git status route). A malicious or malformed cwd silently falls back to the workspace root.
- Not validated / out of scope: legacy (non-workspace-qualified) routes do not get `?cwd=` support, consistent with the existing git status legacy route. Interactive E2E with a real worktree session was not run.
- Breaking changes / migration notes: none — all new parameters are optional; existing callers are unaffected.

## Linked Issues

N/A

<details>
<summary>中文说明</summary>

## 本 PR 做了什么

Web Shell 的 Changes（diff）和 History（log）对话框此前对 worktree session 完全禁用——composer 按钮被隐藏，`/diff` 和 `/log` 斜杠命令始终指向主仓库根目录。本 PR 在整条链路（daemon 路由 → SDK 客户端 → 对话框组件）中传递可选的 git 工作目录参数，使 worktree session 能够查看自身的文件变更和提交历史，与已支持 worktree 的 git 状态指示器行为对齐。

路径包含性校验逻辑（realpath + relative-path 检查）原先内联在 git status 路由中，现提取为共享函数，供 diff 和 log 路由复用。

## 为什么需要

当 session 运行在 git worktree 中时，用户可以看到分支指示器（紫色 fork 图标及 dirty/ahead/behind 计数），但无法打开 Changes 或 History 对话框查看具体变更内容。查看 worktree diff 的唯一途径是通过 `exit_worktree` 工具输出——UI 中不可访问。这使得 worktree session 相比普通 session 体验不完整。

### 如何验证

1. 启动 `qwen serve` 并打开 Web Shell。
2. 创建 worktree session（发送第一条消息前在 git popover 中选择 "Worktree" 模式）。
3. 在 session 中修改一些文件（让 agent 编辑文件）。
4. 点击 composer 工具栏的 git 芯片——Changes 对话框应打开并显示 worktree 的修改文件及正确 diff。
5. 切换到 History 标签——应显示 worktree 分支的提交日志。
6. 验证 `/diff` 和 `/log` 斜杠命令也打开指向 worktree 的对话框。
7. 确认非 worktree session 仍正常工作（Changes/History 指向工作区根目录）。

### 证据（修改前后）

修改前：worktree session 没有 Changes 按钮（被 `!sessionWorktree` 守卫隐藏）；`/diff` 和 `/log` 指向主仓库根目录。
修改后：worktree session 显示 Changes 按钮；对话框通过 `?cwd=` 查询参数指向 worktree 路径。

### 测试环境

|     OS     | 状态 |
| :--------: | :----: |
|  🍏 macOS  |   ✅   |
| 🪟 Windows |   ⚠️   |
|  🐧 Linux  |   ⚠️   |

### 环境（可选）

单元测试：CLI 路由测试（33 通过）、SDK DaemonClient 测试（292 通过）、Web Shell 对话框测试（26 通过）。三个包的类型检查均无新增错误。

## 风险与范围

- 主要风险或权衡：`?cwd=` 参数接受客户端传入的任意路径，但通过 realpath 解析和工作区根目录包含性校验进行验证（与 git status 路由已有的模式一致）。恶意或格式错误的 cwd 会静默回退到工作区根目录。
- 未验证 / 超出范围：legacy（非 workspace-qualified）路由未添加 `?cwd=` 支持，与现有 git status legacy 路由保持一致。未运行真实 worktree session 的交互式 E2E 测试。
- 破坏性变更 / 迁移说明：无——所有新参数均为可选；现有调用方不受影响。

## 关联 Issue

无

</details>

### CODE REVIEW

**Comment by wenshao** [comment]
@qwen-code /takeover

**Comment by wenshao** [comment]
## Review: `fix(web-shell): enable Changes and History dialogs for worktree sessions`

### Overview

Threads an optional git working directory through daemon routes → SDK client → dialog components so worktree sessions can open Changes/History against their own worktree. The path-containment logic that was inlined in the git-status route is extracted into `resolveContainedCwd` and reused by the four diff/log routes.

I diffed the extracted helper against the code it replaced — it is semantically identical (same `realpathSync` on both sides, same `path.relative` containment test, same silent-fallback `catch`), so the refactor half is behavior-preserving. A few things I checked and found correct: both sides are realpath'd so a symlinked `cwd` can't escape; a non-string `cwd` (express/`qs` turns `?cwd=a&cwd=b` into an array and `cwd[x]=y` into an object) falls through the `typeof` guard; `applyReadHeaders` sets `no-store`, so there's no cross-`cwd` cache bleed; echoing the resolved cwd back as the response's `workspaceCwd` matches the precedent the status route already set; effect/memo deps were all updated (`gitCwd` on both content effects and `loadMore`, `sessionWorktree` on the slash-command memo); the `GitDialog` remount key includes `gitCwd`; and the sidebar folder chip correctly does **not** pass `gitCwd`.

---

### Suggestions

**1. The new `?cwd=` / `gitCwd` path has no positive test coverage anywhere — server or client.**

- Server: `workspace-git-diff.test.ts` and `workspace-git-log.test.ts` gained no cases, and grepping `cwd=` across `packages/cli/src/serve/**/*.test.ts` finds nothing for the git routes. The pre-existing inlined version in `workspace-git.ts` was equally untested, so the net effect is that a containment check — the only thing between a client-supplied path and a `git` invocation — is now shared by five routes with zero direct tests. Flipping `!rel.startsWith('..')`, or dropping either `realpathSync`, would keep CI green.
- Client: all four updated assertions assert `undefined`. That's not incidental — `GitDiffDialog.test.tsx` / `GitLogDialog.test.tsx` mount the legacy `GitDiffDialog` / `GitLogDialog` wrappers (`GitDiffDialog.tsx:451`, `GitLogDialog.tsx:365`), which don't accept `gitCwd` and can't be made to pass one. So the parameter is only ever exercised in its inactive state. (Worth noting separately: `GitDialog` is now the sole production consumer of `GitDiffContent`/`GitLogContent`, so those two wrappers are reachable only from their tests.)
- Minimal set I'd want: a unit test for `resolveContainedCwd` (accepts an in-workspace subdir; rejects `../` escape, a symlink resolving outside the root, a non-existent path, an array-valued `cwd`); one route test per file for the honored vs. rejected case; one client test mounting `GitDialog` with `gitCwd` set.

Per AGENTS.md this is a Suggestion rather than a Critical, but of the items here it's the one I'd act on first, because the untested code is the security boundary the PR description leans on.

**2. Silent fallback converts "no data" into confidently-wrong data.**

When containment fails, `resolveContainedCwd` returns the workspace root and the dialog renders the **main checkout's** changes/history with nothing in the UI distinguishing it from the worktree's. Two reachable routes to that state:

- The worktree directory goes away (`exit_worktree`, manual `git worktree remove`) while `sessionWorktree` is still cached client-side — the restoring effect only refetches on `connection.sessionId` change — so `realpathSync` throws and the route silently serves the main repo.
- The workspace root isn't the repo root. `enter_worktree` deliberately resolves to `git rev-parse --show-toplevel` and places worktrees at `<repoRoot>/.qwen/worktrees/<slug>`; for a workspace opened on a repo subdirectory, the sidecar path restored into `session.worktree` (`session.ts` ~2007, `session-list.ts` `enrichWorktreeSidecars`) is outside `runtime.workspaceCwd` and gets rejected. Web-shell-created worktrees use `new GitWorktreeService(workspaceCwd)`, so those are always contained — this only bites the tool-created path.

Before this PR both cases showed no button at all, which is a safer failure. Consider having the helper distinguish "absent" from "rejected" (e.g. return `null` on rejection) so the route can 400 or return `available: false`, and the dialog can say "this worktree is no longer available" instead of rendering a plausible-looking wrong diff.

---

### Nits

- **SDK argument handling is inconsistent across the four methods.** `workspaceGitDiff` gates on truthiness (`cwd ? …`) while the other three use `cwd != null`, so `''` is dropped by one and sent as `cwd=` by the others. Harmless server-side (the route treats `length === 0` as absent), but worth picking one. Separately, the three new call sites use `encodeURIComponent` directly while the file aliases it as `urlEncode` (`DaemonClient.ts:318`) and uses that alias everywhere else, including the adjacent `path`/`oldPath`/`sha` encodings in the same expressions.
- **`workspaceGitLog(limit?, skip?, cwd?)`** is now three positional optionals; an options bag would age better, though positional matches the current file style — take it or leave it.
- **`DiffFileRow` key omits `gitCwd`.** `key={`${workspaceCwd}:${file.path}`}` — the comment above it explains the key is namespaced by workspace precisely to avoid cross-repo path collisions. It's safe today only because `GitDialog`'s key forces a full remount on `gitCwd` change, but `GitDiffContent`'s own effect deps now include `gitCwd`, so its contract allows a `gitCwd` change without a remount, which would leave stale expanded hunks keyed to the old repo. `${workspaceCwd}:${gitCwd ?? ''}:${file.path}` closes it for free.
- **Two `realpathSync` calls per request, synchronously on the event loop** — now on every diff-list, log-list, commit-detail *and* per-file diff request (one per expanded file). The workspace root's realpath is invariant per runtime; caching it on the runtime would remove half of them if this ever shows up in profiles.
- **No worktree indication in the dialog.** The header/subtitle is identical whether you're looking at the workspace or a worktree. Carrying the worktree branch or slug in the subtitle would make #2 self-evident to the user and is a small win regardless.

---

### Verdict

No Critical findings. The refactor is clean and the plumbing is consistent with how the git-status chip already handles worktrees. I'd land it after adding coverage for the containment helper and the honored/rejected `?cwd=` cases (#1), and would give #2 a look since it changes a fail-closed UI into a fail-silent one.

File: packages/web-shell/client/App.tsx [hunk 1]
```diff
@@ -5030,15 +5030,23 @@ export function App({
               pushToast('info', t('localCommand.diffNoWorkspace'));
               return true;
             }
-            setGitDialog({ workspaceCwd: gitDiffWorkspaceCwd, view: 'diff' });
+            setGitDialog({
+              workspaceCwd: gitDiffWorkspaceCwd,
+              gitCwd: sessionWorktree?.path,
+              view: 'diff',
+            });
```
Comment (wenshao) [inline packages/web-shell/client/App.tsx:5071]: **[Critical]** Derive this target from authoritative, session-scoped live cwd state rather than the unversioned `sessionWorktree` snapshot. During a direct switch from worktree session A to session B, `connection.sessionId` changes before `sessionStatus(B)` resolves, so these entry points can still send A's path and display A's diff/history for B. Separately, `POST /session/:id/cd` relocates the live agent but leaves `entry.worktree` unchanged, while Web Shell ignores `session_cwd_changed`, so the dialog keeps targeting the old worktree after a successful cwd change. Associate the metadata with its session id and update/clear the Git target whenever the live cwd changes.

*— Codex GPT-5 via Qwen Code /review*

File: packages/cli/src/serve/workspace-route-runtime.ts [hunk 2]
```diff
@@ -212,3 +213,29 @@ export function sendWorkspaceMismatch(
...
+  if (typeof rawCwd !== 'string' || rawCwd.length === 0) {
+    return workspaceCwd;
+  }
+  try {
+    const resolved = fs.realpathSync(path.resolve(rawCwd));
+    const root = fs.realpathSync(workspaceCwd);
+    const rel = path.relative(root, resolved);
+    if (!rel.startsWith('..') && !path.isAbsolute(rel)) {
+      return resolved;
+    }
```
Comment (wenshao) [inline packages/cli/src/serve/workspace-route-runtime.ts:306]: **[Critical]** Validate a server-owned worktree identity here instead of accepting any contained directory and silently substituting `workspaceCwd` on failure. The current contract both rejects supported repo-top-level worktrees for monorepo-subdirectory workspaces (returning plausible main-checkout data) and accepts a contained directory whose `.git` points at another repository, letting diff/log expose that repository's metadata and removed contents. If a previously accepted worktree disappears between list/detail or pagination requests, the fallback can also mix main-checkout data into an open worktree dialog. Resolve only a registered worktree belonging to this runtime's repository, share the allowed-root logic with session restore, and return a typed 4xx for an explicit invalid or missing `cwd`.

*— Codex GPT-5 via Qwen Code /review*

File: packages/sdk-typescript/src/daemon/DaemonClient.ts [hunk 3]
```diff
@@ -4455,10 +4458,12 @@ export class WorkspaceDaemonClient {
   workspaceGitDiffFile(
     path: string,
     oldPath?: string,
+    cwd?: string,
   ): Promise<DaemonWorkspaceGitDiffHunks> {
```
Comment (wenshao) [inline packages/sdk-typescript/src/daemon/DaemonClient.ts:4776]: **[Suggestion]** Prefer an options object for these optional values. A caller that writes `workspaceGitDiffFile('a.ts', worktreePath)` type-checks but sends the worktree as `oldPath` and omits `cwd`; `workspaceGitLog` similarly requires `undefined, undefined, cwd` when only the target directory is known. Use forms such as `workspaceGitDiffFile(path, { oldPath, cwd })` and `workspaceGitLog({ limit, skip, cwd })`, retaining overloads if compatibility is required.

*— Codex GPT-5 via Qwen Code /review*

**Review by wenshao — COMMENTED** [review]
⚠️ Downgraded from Request changes to Comment: self-PR; CI failing: label. Reviewed. Suggestions are inline. Not reviewed: build-and-test — CLI tests timed out after 300 seconds, and the CI Integration Tests (CLI, No Sandbox) check was skipped. Not reviewed: reverse-audit — stopped at the five-round hard cap after one consecutive dry round.

*— Codex GPT-5 via Qwen Code /review*

File: packages/cli/src/serve/workspace-route-runtime.test.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  });
+
+  it('returns workspaceCwd when cwd is not a string (array)', () => {
+    expect(resolveContainedCwd(fakeReq(['a', 'b']), workspace)).toBe(workspace);
+  });
+
+  it('returns the resolved path for a valid subdirectory', () => {
+    const sub = path.join(workspace, 'sub');
+    fs.mkdirSync(sub);
+    expect(resolveContainedCwd(fakeReq(sub), workspace)).toBe(sub);
```
Comment (doudouOUC) [inline packages/cli/src/serve/workspace-route-runtime.test.ts]: **[Critical]** Two tests fail on macOS because `os.tmpdir()` returns `/var/folders/...` (a symlink to `/private/var/folders/...`), but `resolveContainedCwd` returns `fs.realpathSync`-resolved paths. — Failure scenario: `fs.mkdtempSync(path.join(os.tmpdir(), ...))` produces `/var/folders/...`, while `fs.realpathSync` inside `resolveContainedCwd` resolves to `/private/var/folders/...`. Both `returns the resolved path for a valid subdirectory` (line 47) and `accepts the workspace root itself` (line 51) fail with `expected '/private/var/...' to be '/var/...'`. Verified: 2 failed | 6 passed on macOS. The `test_macos` CI job would also fail. → CI test suite failure on macOS.

```suggestion
    expect(resolveContainedCwd(fakeReq(sub), workspace)).toBe(fs.realpathSync(sub));
```

And similarly for line 51:
```suggestion
    expect(resolveContainedCwd(fakeReq(workspace), workspace)).toBe(fs.realpathSync(workspace));
```

_— qwen3.7-max via Qwen Code /review_

**Review by doudouOUC — DISMISSED** [review]

**Review by doudouOUC — COMMENTED** [review]
Reviewed. 1 Suggestion-level finding(s) could not be anchored to a changed line and were dropped; nothing further to act on here. Unresolved, please confirm: **[Critical]** workspace-route-runtime.ts:236 — Codex blocker: 'validate server-owned worktree identity' — resolveContainedCwd accepts any contained directory; this follows the existing git-status pattern but differs from registry-based validation the blocker requests. Architectural choice, not a clear bug. **[Critical]** App.tsx:5037 — Codex blocker: 'derive target from session-scoped live cwd state' — gitCwd is sourced from sessionWorktree snapshot which updates asynchronously; theoretical stale window during rapid session switching. React key and useEffect deps provide self-correction. Architectural concern, not a clear bug.

_— qwen3.7-max via Qwen Code /review_

File: packages/web-shell/client/components/dialogs/GitLogDialog.test.tsx [hunk 5]
```diff
@@ -308,4 +308,46 @@ describe('GitLogDialog', () => {
...
+        </I18nProvider>,
+      );
+    });
+    await flush();
+
+    expect(workspaceGitLog).toHaveBeenCalledWith(50, 0, '/worktrees/wt');
+
+    const row = document.body.querySelector(
+      'button[aria-expanded="false"]',
+    ) as HTMLButtonElement;
```
Comment (doudouOUC) [inline packages/web-shell/client/components/dialogs/GitLogDialog.test.tsx:354]: **[Suggestion]** The `forwards gitCwd` test covers initial fetch and commit-detail expansion but does not exercise `loadMore` pagination with a non-undefined `gitCwd` — Failure scenario: if `loadMore` accidentally omitted `gitCwd` from the third argument to `workspaceGitLog()`, all existing tests would still pass because the pagination test only asserts `undefined` as the third arg, and this test never clicks "Load more". The analogous file-expansion path in `GitDiffContent` IS covered.

```suggestion
    expect(workspaceGitLog).toHaveBeenCalledWith(50, 0, '/worktrees/wt');

    // Click 'Load more' to verify gitCwd is forwarded through pagination
    const loadMore = document.body.querySelector(
      'button.git-load-more',
    ) as HTMLButtonElement;
    if (loadMore) {
      fireEvent.click(loadMore);
      expect(workspaceGitLog).toHaveBeenNthCalledWith(2, 50, 1, '/worktrees/wt');
    }

    const row = document.body.querySelector(
      'button[aria-expanded="false"]',
    ) as HTMLButtonElement;
```

_— qwen3.7-max via Qwen Code /review_

**Review by doudouOUC — COMMENTED** [review]
Reviewed. Suggestions are inline. Unresolved, please confirm: **[Critical]** workspace-route-runtime.ts:236 — Codex blocker: 'validate server-owned worktree identity' — containment-based validation (not registry-based); pre-existing architectural pattern, not a defect introduced by this PR **[Critical]** App.tsx:5039 — Codex blocker: 'derive target from session-scoped live cwd state' — theoretical stale window during session switching; React key + useEffect deps provide self-correction; pre-existing architectural concern

_— qwen3.7-max via Qwen Code /review_

**Comment by ytahdn** [comment]
## 🔍 Code Review

**13 files reviewed** · 0 findings · LGTM

Clean parameter-threading change: extracts `resolveContainedCwd()` from the git-status route into a shared helper (same realpath + containment logic, no new attack surface), then threads an optional `gitCwd` through daemon routes → SDK client → dialog components so worktree sessions can view their own diffs and history.

Verified: path traversal safety (realpath + relative-path containment), URL encoding (`urlEncode` / `URLSearchParams`), useEffect dependency arrays include `gitCwd`, key props prevent stale state on worktree switch, fallback to workspace root on invalid cwd, and backward compatibility (all new params optional).

*Reviewed with manual audit (OCR unavailable — GitHub git protocol timeout)*

**Review by ytahdn — DISMISSED** [review]
LGTM. Clean cwd threading for worktree diff/log, shared containment helper, no new attack surface. 0 findings.

**Review by doudouOUC — COMMENTED** [review]
Reviewed. Suggestions are inline. Unresolved, please confirm: **[Critical]** workspace-route-runtime.ts:236 — Codex blocker: 'validate server-owned worktree identity' — containment-based validation is a pre-existing architectural pattern, not a defect introduced by this PR **[Critical]** App.tsx:5039 — Codex blocker: 'derive target from session-scoped live cwd state' — theoretical timing concern with self-correcting React mechanisms (key + useEffect deps), pre-existing architectural concern

_— qwen3.7-max via Qwen Code /review_

File: packages/sdk-typescript/src/daemon/DaemonClient.ts [hunk 6]
```diff
@@ -4737,10 +4737,12 @@ export class WorkspaceDaemonClient {
     );
   }
 
-  workspaceGitDiff(): Promise<DaemonWorkspaceGitDiff> {
+  workspaceGitDiff(cwd?: string): Promise<DaemonWorkspaceGitDiff> {
+    const suffix =
+      cwd != null ? `/git/diff?cwd=${urlEncode(cwd)}` : '/git/diff';
```
Comment (doudouOUC) [inline packages/sdk-typescript/src/daemon/DaemonClient.ts:4763]: **[Suggestion]** SDK `cwd` URL-encoding path has no unit test coverage — Failure scenario: a typo in the query key (e.g. `Cwd=` instead of `cwd=`), a missing `urlEncode` call, or an incorrect `&`/`?` separator in any of the four updated methods would silently produce the wrong URL. The route would ignore the unrecognized parameter and return workspace-root data instead of worktree data, and no test would catch it.

Consider adding tests that call each of the four methods (`workspaceGitDiff`, `workspaceGitDiffFile`, `workspaceGitLog`, `workspaceGitCommitDetail`) with a concrete `cwd` value and assert the exact resulting URL.

_— qwen3.7-max via Qwen Code /review_

File: packages/cli/src/serve/routes/workspace-git-diff.ts [hunk 7]
```diff
@@ -200,7 +201,7 @@ export function registerWorkspaceQualifiedGitDiffRoutes(
     if (!runtime) return;
     void handleDiffList(
       res,
-      runtime.workspaceCwd,
+      resolveContainedCwd(req, runtime.workspaceCwd),
       deps.sendBridgeError,
       'GET /workspaces/:workspace/git/diff',
```
Comment (doudouOUC) [inline packages/cli/src/serve/routes/workspace-git-diff.ts:227]: **[Suggestion]** Route-level tests for the `?cwd=` query parameter integration are missing — Failure scenario: if a route accidentally passes `req.query['cwd']` to a different argument position, or if `resolveContainedCwd` is called with the wrong `req`/`runtime.workspaceCwd`, the route would silently operate on the workspace root instead of the requested worktree. The `resolveContainedCwd` unit tests cover the pure function, but not the HTTP-layer integration.

Consider adding route-level tests that send requests with `?cwd=<contained-path>` and assert the downstream handler receives the resolved path, plus a test where `cwd=` escapes the workspace and the route falls back to `workspaceCwd`. Apply to all four affected routes (diff list, diff file, log list, commit detail).

_— qwen3.7-max via Qwen Code /review_

**Comment by wenshao** [comment]
@qwen-code /triage

File: packages/web-shell/client/App.tsx [hunk 8]
```diff
@@ -7889,10 +7899,11 @@ export function App({
                           onGitModeIntentChange={gitModeEligible ? setGitModeIntent : undefined}
                           gitStatus={selectedWorkspaceGitStatus}
                           onOpenGitDiff={
-                            gitDiffWorkspaceCwd && !sessionWorktree
+                            gitDiffWorkspaceCwd
```
Comment (yiliang114) [inline packages/web-shell/client/App.tsx:7934]: **[P2] Confirm: Changes button now opens for worktree sessions — intentional UX change**

The guard changed from `gitDiffWorkspaceCwd && !sessionWorktree` to just `gitDiffWorkspaceCwd`. This means the composer Changes button is now visible for worktree sessions (previously hidden). Combined with `gitCwd: sessionWorktree?.path`, the dialog will target the worktree.

Looks intentional per the PR title, but worth confirming this is the desired UX — previously the button was hidden because `/diff` always targeted the main workspace root. Now that cwd threading works, showing the button makes sense.

No action needed if this is deliberate.

**Review by yiliang114 — APPROVED** [review]
LGTM. Clean, well-structured PR that correctly threads the worktree cwd through the full stack.

**Summary of review:**

1. **Containment check** (`resolveContainedCwd`): Solid. Extracted from inline duplication in `workspace-git.ts` into a shared helper. The `realpathSync` + `path.relative` + `startsWith("..")` pattern correctly handles symlink escape and path traversal. Test suite covers all edge cases (symlinks, outside paths, missing paths, non-string inputs).

2. **Threading correctness**: `gitCwd` flows properly: `sessionWorktree?.path` → `gitDialog` state → `GitDialog` → `GitDiffContent`/`GitLogContent` → SDK calls → daemon routes → `resolveContainedCwd`. All React hooks have correct dependency arrays including `gitCwd`.

3. **Backward compat**: All new SDK parameters are optional with `undefined` fallback. Routes default to `runtime.workspaceCwd` when `?cwd=` is absent. Existing tests updated to pass explicit `undefined`.

4. **Remounting**: Dialog `key` now includes `gitCwd` (`${workspaceCwd}:${gitCwd ?? }:${view}`), ensuring fresh state when switching worktrees. Diff file rows also include `gitCwd` in their key.

5. **One P2 observation** (inline): The composer Changes button guard dropped `!sessionWorktree` — this is the intended UX change per the PR title, just confirming.

No P0/P1 blockers found.

**Review by yiliang114 — DISMISSED** [review]
**Approve** ✅

Clean worktree path threading through all 5 layers (sessionWorktree → dialog state → SDK → daemon route → resolveContainedCwd). Security is solid — `realpathSync` + `path.relative` + `startsWith('..')` containment check properly rejects outside paths and symlink escapes. All new SDK params are optional, routes fall back to workspaceCwd. 8 dedicated containment tests + worktree forwarding tests in all dialog components.

**Comment by wenshao** [comment]
@qwen-code /resolve

**Comment by wenshao** [comment]
@qwen-code /triage

**Review by ytahdn — APPROVED** [review]
Reviewed at e44ce92dc516abcfa538d38a88bfb7fa79a71b51.
